enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,291 @@
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If a CSV defines a upgradeReadiness Service without a spec that does not exist, upgrades for that operator are blocked.
+- If a CSVless bundle defines a upgradeReadiness Service that does not exist, the bundle fails to build.
+- If a CSV defines an upgradeReainess Service with a spec, changes to the service are reverted by OLM.
+- If a CSV/CSVless bundle defines an upgradeReainess Service that returns a 0, upgrades to the operator are blocked. The operator then upgrades when the service is no longer blocked.
+
+Additionally, tests that use CSVs that omit the upgradeReainess Service must continue to adhear to existing behavior.
+
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Alternatives
+
+The are viable alternatives to the solution proposed in this enhancement.
+
+### Subsbitute the UpgradeReadinessService with an UpdateProbe
+
+It is possible that we could take advantage of the standard defined by [Liveness, Readiness, and StartUp Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) by importing the [Prober](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/kubelet/prober/prober.go#L48) code. There are cons and benefits to this approach defined below:
+
+**This seems to be the most viable alternative.**
+
+#### Pros
+
+- Follows a well defined probe format which supports [HTTP Gets, EXEC, and TCPSocket](https://github.com/kubernetes/api/blob/6f652b6ce59c386f4a431eb031d0339620aaff5e/core/v1/types.go#L2279-L2292) options.
+- Enables OLM to identify which POD is reporting a upgradeReadiness failure, as a service could route to one of any number of pods.
+
+#### Cons
+- Requires OLM create an import and maintain prober code which executes the probe.
+
+### Create an API so Operators may report Status
+
+OLM could introduce a new API that allows operators to report their upgradeReadiness
+
+#### Pros
+
+- Provides a single source of truth for upgradeReadiness.
+
+#### Cons
+
+- Increases the number of APIs that OLM introduces. OLM has already received negative feedback on the number of APIs one must work with.
+- Behaves differently from other standard Kubernetes probes.
+
+### Operators write to the status of the CSV/operator API
+
+OLM could allow operator authors to write to the status object of an API that OLM manages.
+
+#### Pros
+
+- No new APIs
+- Should require less OLM code
+
+#### Cons
+
+- Places ownice on Operator Authors to learn OLM specific readiness reporting conventions.
+- Requires Operator Authors to write to the status of an OLM owned resource.
+- OLM would probably need to provide operator authors with a library to report readiness to standardize the approach.
+- Operators could place the operator's status in a bad state.
```

Comment:
What do you mean by this?

Reply:
With this approach, operator author are now responsible for editing the status of the CSV or Operator object. OLM has no ability to limit the operator from changing items other than readiness within the status. For example, imagine an operator that accidentally removes the phase from the CSV Status. OLM would kick off an install.